symengine/serialize-cereal.h

```diff
@@ -0,0 +1,223 @@
+    case type:                                                                 \
+        save_basic(ar, static_cast<const Class &>(*ptr));                      \
+        break;
+#include "symengine/type_codes.inc"
+#undef SYMENGINE_ENUM
+            default:
+                save_basic(ar, *ptr);
+        }
+    }
+}
+
+//! Saving for SymEngine::RCP
+template <class Archive, class T>
+inline void CEREAL_SAVE_FUNCTION_NAME(Archive &ar, RCP<const T> const &ptr)
+{
+    save_basic(ar, rcp_static_cast<const Basic>(ptr));
+}
+
+template <class Archive>
+RCP<const Basic> load_basic(Archive &ar, RCP<const Symbol> &)
+{
+    std::string name;
+    ar(name);
+    return symbol(name);
+}
+template <class Archive>
+RCP<const Basic> load_basic(Archive &ar, RCP<const Pow> &)
+{
+    RCP<const Basic> base, exp;
+    ar(base);
+    ar(exp);
+    return make_rcp<const Pow>(base, exp);
+}
+template <class Archive>
+RCP<const Basic> load_basic(Archive &ar, RCP<const Add> &)
+{
+    RCP<const Number> coeff;
+    umap_basic_num dict;
+    ar(coeff);
+    ar(dict);
+    return make_rcp<const Add>(coeff, std::move(dict));
+}
+template <class Archive>
+RCP<const Basic> load_basic(Archive &ar, RCP<const Integer> &)
+{
+    std::string name;
+    ar(name);
+    return integer(integer_class(name));
+}
+// template <class Archive>
+// RCP<const Basic> load_basic(Archive &ar, RCP<const Rational> &)
+// {
+//     RCP<const Integer> num;
+//     RCP<const Integer> den;
+//     ar(num, den);
+//     return Rational::from_two_ints(*num, *den);
+// }
+
+#define SYMENGINE_ENUM(type, Class)                             \
+    template <class Archive>                                    \
+    RCP<const Basic> load_basic(Archive &ar, RCP<const Class> &)\
+    {                                                           \
+        RCP<const Number> arg;                                  \
+        ar(arg);                                                \
+        return make_rcp<const Class>(arg);                      \
+    }
+#include "symengine/type_codes_oneargfunction.inc"
+#undef SYMENGINE_ENUM
+
+#define SYMENGINE_ENUM(type, Class)                                     \
+    template <class Archive>                                            \
+    RCP<const Basic> load_basic(Archive &ar, RCP<const Class> &)        \
+    {                                                                   \
+        RCP<const Number> arg1;                                         \
+        RCP<const Number> arg2;                                         \
+        ar(arg1, arg2);                                                 \
+        return make_rcp<const Class>(arg1, arg2);                       \
+    }
+#include "symengine/type_codes_boolean_relational.inc"
+#undef SYMENGINE_ENUM
+
+
+template <class Archive, typename T>
+RCP<const Basic> load_basic(Archive &ar, RCP<const T> &b)
+{
+    const auto t_code = b->get_type_code();
```

Comment:
`b` is a dummy to allow writing code with inheritance. Don't use it.

Reply:
Sure, but isn't it guaranteed that it inherits from Basic?

Reply:
Yes, but the pointer inside might be NULL, so you can't call methods on it.